### PR TITLE
Implement aten.select.int sharding strategy

### PR DESCRIPTION
**Pull request by kkkkeeee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149842

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @tianyu-l @XilunWu

### CODE REVIEW

**Review by jingsh — CHANGES_REQUESTED** [review]
suggest split the diff to 2, one for auto formatting and one for the implementation

**Review by tianyu-l — COMMENTED** [review]
Thanks for the PR. Would you please add test cases, covering different input sharding possibilities?

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 1]
```diff
@@ -236,6 +236,81 @@ def gen_bucketize_strategy(op_schema: OpSchema) -> StrategyType:
...
+@register_op_strategy(aten.select.int, schema_info=RuntimeSchemaInfo(1))
+def gen_select_strategy(op_schema: OpSchema) -> StrategyType:
+    """
+    It contains three cases:
+    1. input is replicated, output is replicated
+    2. input is sharded, and shard_dim == selected_dim, first replicate the tensor
+       across the mesh, and do the select. The output is replicated.
+    3. input is sharded, and shard_dim != selected_dim, the output is sharded. Need to
+       figure out the output_shard_dim.
+    """
```
Comment (XilunWu) [inline torch/distributed/tensor/_ops/_tensor_ops.py:248]: The comment is very helpful in reader's understanding. One next step is to add support for `Partial` placement. Let's sync offline to go through this.

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 2]
```diff
@@ -236,6 +236,81 @@ def gen_bucketize_strategy(op_schema: OpSchema) -> StrategyType:
...
+    )
+    input_shape = input_strategy.shape
+    input_ndim = input_strategy.ndim
+    selected_dim = normalize_dim(selected_dim, input_ndim)
+    index = normalize_dim(index, input_shape[selected_dim])
+
+    select_strategy = OpStrategy([])
+    for arg_strategy in input_strategy.strategies:
+        arg_spec = arg_strategy.output_spec
+        placements = arg_spec.placements
```
Comment (XilunWu) [inline torch/distributed/tensor/_ops/_tensor_ops.py:264]: this `placements`  is unused, isn't it?

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 3]
```diff
@@ -236,6 +236,81 @@ def gen_bucketize_strategy(op_schema: OpSchema) -> StrategyType:
...
+
+        if arg_spec.is_sharded():
+            # shard_dim == selected_dim, R() -> R()
+            if is_tensor_dim_sharded(arg_spec, dim=selected_dim):
+                # if the input is sharded on the split dim, we need to unshard it
+                placements = replicate_tensor_dim(arg_spec.placements, dim=selected_dim)
```
Comment (XilunWu) [inline torch/distributed/tensor/_ops/_tensor_ops.py:270]: using `replicate_tensor_dim` means we're also tackling `Partial` placement. If you intend to use it, let's clarify that in the above comment as well.

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 4]
```diff
@@ -236,6 +236,81 @@ def gen_bucketize_strategy(op_schema: OpSchema) -> StrategyType:
...
+
+                output_shard_dim = None
+                for i in range(len(tensor_shape)):
+                    if tensor_shape[i] == 1:
+                        output_shard_dim = i
+                        break
+                assert output_shard_dim is not None
+                output_specs = DTensorSpec(
+                    arg_spec.mesh, placements=(Shard(dim=output_shard_dim),)
+                )  # Shard
```
Comment (XilunWu) [inline torch/distributed/tensor/_ops/_tensor_ops.py:298]: I could be wrong - initially I think the case `shard_dim != select_dim` is quite simple, you just keep the original input spec without doing any redistribution. For example,
```
x  # shape=[4, 4], placements=[S(1)]
x[0]  # still has placements=[S(1)] and no redistribution is required to perform this select op
```

Is there anything I'm missing?

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 5]
```diff
@@ -236,6 +236,81 @@ def gen_bucketize_strategy(op_schema: OpSchema) -> StrategyType:
...
+
+        else:
+            # input_spec is R, output_spec is R
+            input_specs = arg_spec
+            output_specs = arg_spec
```
Comment (XilunWu) [inline torch/distributed/tensor/_ops/_tensor_ops.py:303]: If my understanding above is correct, this branch can be merged

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 4 above)
Comment (kkkkeeee) [reply torch/distributed/tensor/_ops/_tensor_ops.py:298]: When the tensor shape is 2D, x.shape=[4, 4] and placements=[S(1)], if mesh=[0,1], the above is true. If mesh=[0, 1, 2, 3], then the shard_dim may change.

When mesh=[0, 1, 2, 3], x.shape=[4, 4], placements=[S(1)], , select_dim=0, each shard holds a [4, 1] tensor. If we do a selection op on dim=0, the returned value is a 1D tensor. So the placements becomes S(0).

Similarly, when the tensor shape is 3D, we have the following cases with mesh=[0, 1, 2, 3]:
Case 1, x.shape=[4, 4, 4], placements=[S(0)], select_dim=1  # output placements=[S[0]]
Case 2, x.shape=[4, 4, 4], placements=[S(0)], select_dim=2  # output placements=[S[0]]
Case 3, x.shape=[4, 4, 4], placements=[S(1)], select_dim=0  # output placements=[S[0]], *not the same as input S(1)*
Case 4, x.shape=[4, 4, 4], placements=[S(1)], select_dim=2  # output placements=[S[1]]
Case 5, x.shape=[4, 4, 4], placements=[S(2)], select_dim=0  # output placements=[S[1]], *not the same as input S(2)*
Case 5, x.shape=[4, 4, 4], placements=[S(2)], select_dim=1  # output placements=[S[1]], *not the same as input S(2)*

I think the above code only handles the case with len(x[shard_dim]) = len(mesh) case. I need to add another case when len(x[shard_dim]) > len(mesh), which is what you mentioned case.

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 5 above)
Comment (kkkkeeee) [reply torch/distributed/tensor/_ops/_tensor_ops.py:303]: This is for the branch when input tensor is not sharded. I think it is better we keep this branch?

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 2 above)
Comment (kkkkeeee) [reply torch/distributed/tensor/_ops/_tensor_ops.py:264]: Good catch, will remove it

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -7647,6 +7647,7 @@ def sample_inputs_select(op_info, device, dtype, requires_grad, **kwargs):
              ((S, S, S), (-1, 2)),
              ((S, S, S), (-1, -1)),
              ((S, S, S), (1, -1)),
+             ((S, S), (-1, 2)),
```
Comment (XilunWu) [inline torch/testing/_internal/common_methods_invocations.py:7650]: Why do we add this line?

**Review by XilunWu — APPROVED** [review]
LGTM! One question.

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (kkkkeeee) [reply torch/testing/_internal/common_methods_invocations.py:7650]: It seems we don't have any test on 2D tensor. So I just added one.

Comment (XilunWu) [reply torch/testing/_internal/common_methods_invocations.py:7650]: hmm... Isn't the 3D tensor case covering the 2D? 

if `((S, S), (-1, 2))` fails, `((S, S, S), (-1, 2))` would also fail.

Comment (kkkkeeee) [reply torch/testing/_internal/common_methods_invocations.py:7650]: Just think it doesn't hurt to add a 2D use case. We also have 1D use case in the below `((S,), (0, 2))`

Comment (XilunWu) [reply torch/testing/_internal/common_methods_invocations.py:7650]: I see. This makes sense, thanks for clarification!

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 1 above)
Comment (tianyu-l) [reply torch/distributed/tensor/_ops/_tensor_ops.py:251]: @XilunWu Could you clarify a bit on what needs to be done for `Partial`? Currently the `Partial` would just flow through from input to output, which seems right to me.
